libcnb/src/layer/handling.rs

```diff
@@ -0,0 +1,679 @@
+                        .ok_or(HandleLayerError::UnexpectedMissingLayer)
+                        .map_err(HandleLayerErrorOrBuildpackError::HandleLayerError)
+                }
+            }
+        }
+        Err(ReadLayerError::LayerContentMetadataParseError(_)) => {
+            match read_layer::<GenericMetadata, _, _>(&context.layers_dir, layer_name.as_ref()) {
+                Ok(Some(generic_layer_data)) => {
+                    let metadata_migration_strategy = layer
+                        .migrate_incompatible_metadata(
+                            context,
+                            &generic_layer_data.content_metadata.metadata,
+                        )
+                        .map_err(HandleLayerErrorOrBuildpackError::BuildpackError)?;
+
+                    match metadata_migration_strategy {
+                        MetadataMigration::RecreateLayer => {
+                            delete_layer(&context.layers_dir, layer_name.as_ref())?;
+                        }
+                        MetadataMigration::ReplaceMetadata(migrated_metadata) => {
+                            write_layer(
+                                &context.layers_dir,
+                                layer_name.as_ref(),
+                                generic_layer_data.env,
+                                LayerContentMetadata {
+                                    types: generic_layer_data.content_metadata.types,
+                                    metadata: migrated_metadata,
+                                },
+                            )?;
+                        }
+                    }
+
+                    handle_layer(context, layer_name, layer)
+                }
+                Ok(None) => Err(HandleLayerError::UnexpectedMissingLayer.into()),
+                Err(read_layer_error) => {
+                    Err(HandleLayerError::ReadLayerError(read_layer_error).into())
+                }
+            }
+        }
+        Err(read_layer_error) => Err(HandleLayerError::ReadLayerError(read_layer_error).into()),
+    }
+}
+
+fn handle_create_layer<B: Buildpack + ?Sized, L: Layer<Buildpack = B>>(
+    context: &BuildContext<B>,
+    layer_name: impl AsRef<str>,
+    layer: L,
+) -> Result<LayerData<L::Metadata>, HandleLayerErrorOrBuildpackError<B::Error>> {
+    let layer_dir = context.layers_dir.join(layer_name.as_ref());
+
+    let layer_result = layer
+        .create(context, &layer_dir)
+        .map_err(HandleLayerErrorOrBuildpackError::BuildpackError)?;
+
+    write_layer(
+        &context.layers_dir,
+        layer_name.as_ref(),
+        layer_result.env.unwrap_or_default(),
+        LayerContentMetadata {
+            types: layer.types(),
+            metadata: layer_result.metadata,
+        },
+    )?;
+
+    read_layer(&context.layers_dir, layer_name.as_ref())?
+        .ok_or(HandleLayerError::UnexpectedMissingLayer)
+        .map_err(HandleLayerErrorOrBuildpackError::HandleLayerError)
```

Comment:
What conditions will cause write to pass but read to fail? On the surface without more context, this pattern seems overly defensive. I feel like I'm missing some context.

Also we've got this same logic repeated 3 times (handle_create_layer, handle_update_layer, and in ExistingLayerStrategy::Keep condition above). Would this benefit from an extraction? Something like `fn write_then_read` ?

Reply:
> What conditions will cause write to pass but read to fail? On the surface without more context, this pattern seems overly defensive. 

"Because `read_layer` returns a `Result`" is the "technically correct" answer. We really have no option to not handle these cases, and that's great! I agree that this will only happen in extreme cases, but _it can happen_. Because Rust is forcing us to handle these cases, we got a more robust program. Sure, we could call `.unwrap()`, causing panics in these cases, but what would we get out of it? The code here would be barely shorter but much much more unsafe.

To answer your question more concretely: Other programs might modify disk contents while our program is running, it's a race condition. Our program does not have an open transaction to the file system. The hardware might fail between the `write` and `read` calls. The directory could be mounted as write only, etc.

Reply:
> Also we've got this same logic repeated 3 times

I found the cases you mention. I don't think it's worth factoring out at this stage. We'd remove the following three times: 

```rust
        .ok_or(HandleLayerError::UnexpectedMissingLayer)
        .map_err(HandleLayerErrorOrBuildpackError::HandleLayerError)
```

but will have to add a new function that has the same complex signature as `write_layer`. If we're just looking a LOC, we will end up with more code than before.

Also, since `read_layer` and `write_layer` have different error types, we would need another disjoint union type to allow returning either `ReadError` or `WriteError` from that new function, adding more code.

What do we gain? No duplication of the 
```rust
        .ok_or(HandleLayerError::UnexpectedMissingLayer)
        .map_err(HandleLayerErrorOrBuildpackError::HandleLayerError)
```

code. I generally agree that duplication is bad, but this code has no cyclomatic complexity and the only thing that could really go wrong is mismatched types which will be caught by the compiler.

Reply:
Thanks